Implementation of sample selection estimators

mychaelka opened the pull request:
### Description
This PR contains an implementation of two estimators of sample selection models from _Michela Bia, Martin Huber & Lukáš Lafférs (2023) Double Machine Learning for Sample Selection Models, Journal of Business & Economic Statistics, DOI: [10.1080/07350015.2023.2271071](https://doi.org/10.1080/07350015.2023.2271071)_ -- identification under missingness at random and under nonignorable nonresponse, along with basic tests on simulated data. For testing, the file `conftest.py` was also modified to include the DGP for these models. Original implementation of these estimators is available in the R `causalweight` package (https://cran.r-project.org/web/packages/causalweight/index.html).

### Reference to Issues or PRs
None

### Comments
These estimators require a sample selection indicator to be present in the data (1 if outcome is observed, 0 otherwise). The `DoubleMLData` interface does not have a selection indicator available yet, so the implementation uses the time indicator  `t` in its place. The third estimator in the paper (identification under sequential conditional independence) is not implemented yet, as it would require interfering with the implementation of `DoubleMLData`, as it requires the covariates to be split into two parts -- observed pre-treatment and observed post-treatment. 

### PR Checklist
Please fill out this PR checklist (see our [contributing guidelines](https://github.com/DoubleML/doubleml-for-py/blob/main/CONTRIBUTING.md#checklist-for-pull-requests-pr) for details).

- [x] The title of the pull request summarizes the changes made.
- [x] The PR contains a detailed description of all changes and additions.
- [x] References to related issues or PRs are added.
- [x] The code passes all (unit) tests.
- [x] Enhancements or new feature are equipped with unit tests.
- [x] The changes adhere to the PEP8 standards.

SvenKlaassen commented:
Thank you very much for the contribution to the package. 

I think the idea with using the time variable `t` as a sample selection indicator is fine. 
Maybe we can move the PR to a new branch and I can implement an optional argument in the `DoubleMLData` class, before we merge this onto `main`.

mychaelka commented:
> Thank you very much for the contribution to the package.
> 
> I think the idea with using the time variable `t` as a sample selection indicator is fine. Maybe we can move the PR to a new branch and I can implement an optional argument in the `DoubleMLData` class, before we merge this onto `main`.

Thank you Sven, there still remains one estimator in the Sample selection paper (identification under sequential conditional independence) that I have not yet implemented, as mentioned in the comments. This estimator accounts for the covariates X being observed both pre-treatment and post-treatment. Would it be possible to also add such (optional) distinction into `DoubleMLData`?

SvenKlaassen commented:
I guess it would be possible. I will check out the paper and try to come up with a solution.
Today, I did merge the PR to restructure the package. I would like to include the sample selection models in a seperate folder (like irm, did etc.). Maybe you have a good suggestion for a short name?

mychaelka commented:
I was actually thinking about that, since I don't really like the class name that I am using now (`DoubleMLS`), so maybe we could change it to `DoubleMLSEL`, and the short folder name could be "sel"? What do you think?

SvenKlaassen commented:
Both `DoubleMLSEL` and `sel` are fine for me.
Another option would be `ssm` for sample selection model.

I have thought a bit about the additional implementation of covariates $X$ and $M$ for identification under sequential conditional independence and think this would fit better with a larger rewrite of the `DoubleMLData` class.
I was already thinking about an option to distinguish between covariates which are confounders and covariates which only affect the outcome. 
My sugesstion would be to focus on the implmentation of missing at random at nonignorable and in an additional branch i try to extend `DoubleMLData` to be able to handle multiple categories of covariates. But this might take some time. 
On the branch for the current selection models i will add an option `s` for the sample selection indicator similar to `t` for DiD.

mychaelka commented:
`ssm` also sounds good. I agree that we can for now focus on the first two estimators and leave the identification under sequential conditional independence until you have time to extend `DoubleMLData`. I will have some time to work at it this weekend, so I will look at the issues found by CodeQL and at the open points you mentioned in email (exception tests, example notebooks, etc.)

mychaelka commented:
@SvenKlaassen I added model defaults and return type tests for the sample selection models into the existing files and created a new file with exception tests. I am also working on example notebooks with both simulated and real data.
I can see that there is some issue with Codacy analysis, but I'm not sure whether I am able to do something about this?

SvenKlaassen commented:
Sorry for the late reply. 
Thank you for all the updates.
I can resolve the codacy issue later (or after we merge this onto the DoubleML branch).
Can you try to answer my comments in if you changed something. That would be a bit easier for me to retrace the changes.
Especially the comment regarding the ordering and variance estimation?
This is not changed yet right? This might also help with the results on the example notebooks.

mychaelka commented:
Sorry, I couldn't see any comments on the code, nor did I receive any notification. Until now I thought you had been busy and did not have time to go through the code, so I was working on the things you mentioned in email. I still cannot see any comments on the though...

SvenKlaassen commented:
I mean the comments in this PR. I am not sure why you are not able to see them.
We could also do a short call and discuss any open points.

mychaelka commented:
I can only see the comments in this conversation. I tried looking under changed files and commits, but I can only see the warnings from Codacy. We can definitely do a short call -- I am currently away but will be available today around 4pm again. I will also be available during the weekend and for the most of next week.

SvenKlaassen commented:
Sorry, it was completely my fault. I forgot to submit the review....
Not all comments are still valid. 

If some points need discussion, we can still talk (just ask via mail).

SvenKlaassen commented on doubleml/double_ml_selection.py:
I think this is not relevant anymore. We should go foreward with the current implementation.

mychaelka commented:
Thank you, I will try to go over the comments during this weekend. I can already see that some of the comments might be a huge help for the parts that I was struggling with a bit.

mychaelka commented on doubleml/double_ml_selection.py:
```diff
@@ -0,0 +1,607 @@
+            y_0 = np.hstack(y_0)
+            dcontrol = np.hstack(dcontrol)
+
+        p_hat_d0['preds'] = _trimm(p_hat_d0['preds'], self._trimming_rule, self._trimming_threshold)
```
Trimming is done a bit differently in the sample selection paper (see Appendix E in supplementary materials: https://www.tandfonline.com/doi/suppl/10.1080/07350015.2023.2271071?scroll=top). 
It does not truncate the values but rather drops observations for which the product of the treatment propensity score and the selection propensity score is lower than given threshold. Since such option is not supported by DoubleML, I used the truncating option for now, but I guess I could apply this rule to the product of `pi_hat` and `p_hat` instead of just `p_hat`. 

Moreover, the original R implementation `causalweight` has the option to calculate ATE just for the selected subpopulation and in that case, only `p_hat` is trimmed, but I have not implemented this option yet.

SvenKlaassen replied:
But generally the clipping/truncating of values would also be an option right?

mychaelka replied:
Yes, it should be fine to use it.

mychaelka commented on doubleml/double_ml_selection.py:
```diff
@@ -0,0 +1,607 @@
+        return res
+
+    def _sensitivity_element_est(self, preds):
+        # TODO: RR calculation needs to be finished
```
That would be great. I don't have much background in functional analysis and didn't know how to calculate the Riesz representer properly.

SvenKlaassen replied:
I think best option would be to finish everything else and then add this seperately.

mychaelka commented on doubleml/double_ml_selection.py:
```diff
@@ -0,0 +1,607 @@
+            psi_b1 = (dtreat * s_1 * (y_1 - mu_d1)) / (p_d1 * pi_d1) + mu_d1
+            psi_b0 = (dcontrol * s_0 * (y_0 - mu_d0)) / (p_d0 * pi_d0) + mu_d0
+
+        psi_b = psi_b1 - psi_b0
```
You are absolutely right, I don't know why I didn't notice this. I asked Lukas Laffers about using just one sample splitting and he confirmed that it should be OK, so I will rewrite the code to use just one procedure and that should also help me to eliminate this issue. I will also ask about the variance estimation.

mychaelka replied:
Yes, I already started rewriting the code so that it would all match the order of the input data. I think I should be able to get back to it this week.

mychaelka commented:
@SvenKlaassen I refactored the code to use only one sample splitting procedure for the nested estimation. I also fixed the ordering -- now the order of predictions should match the input data. If you come across any other issue (or find the refactored code to still have bugs), please let me know.

SvenKlaassen commented on doubleml/double_ml_ssm.py:
```diff
@@ -0,0 +1,476 @@
+
+    normalize_ipw : bool
+    Indicates whether the inverse probability weights are normalized.
+    Default is ``True``.
```
This is ``False``.

SvenKlaassen commented on doubleml/double_ml_ssm.py:
```diff
@@ -0,0 +1,476 @@
+                dx_train_1 = dx[train_inds_1, :]
+
+                # preliminary propensity score for selection
+                ml_pi_prelim = clone(self._learner['ml_pi'])
```
These models are not saved right? I guess the ``pi_hat['models']`` are ``None``. (see e.g. https://github.com/mychaelka/doubleml-for-py/blob/fb6a38fd7bbe5632fda7473b48e333e224b5a22e/doubleml/double_ml_pq.py#L299)
If i am reading your [paper](https://arxiv.org/abs/2012.00745) correctly, these models are not refitted right? Then we might already remove the preliminary part of the name (this was just used in Quantile models for a 2-step procedure since the propensity score has to be refitted).

mychaelka replied:
You're right, I set the models to be saved (hopefully it's correct this way) and removed the preliminary part of the name (I used it to emphasize that nuisance function pi is estimated first on a different subsample). I also set the above and below mentioned defaults to False.

SvenKlaassen replied:
Thank you. Looks fine to me. I will check it in detail later.

SvenKlaassen commented on doubleml/double_ml_ssm.py:
```diff
@@ -0,0 +1,476 @@
+                         apply_cross_fitting)
+
+        self._external_predictions_implemented = False
+        self._sensitivity_implemented = True
```
Better set this to ``False``

SvenKlaassen reviewed (changes requested):
Looks really good. Only very small comments.
If the models are included, we can merge this and I can then start to update the documentation.
I might also start to add external predictions and sensitivity analysis before merging in onto ``main``.

SvenKlaassen commented:
I will merge the PR onto the extra branch and try to merge it onto `main` later. 
This might take some time as I am quite busy right now and have to previously include some changes to align the implementation with the new framework setting.
I will also update the documentation, so I might ask you in the future to check if you are satisfied with the description (If you have a notebook for the example section, I will happily include that).

mychaelka commented:
Of course, thank you. If you need any help with the documentation or anything else, please let me know. I already have some example notebooks but I have to clean them up first (so far they have been only experimental).
